internal/pkg/prune/prune.go

```diff
@@ -0,0 +1,133 @@
+limitations under the License.
+*/
+
+package prune
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"gopkg.in/src-d/go-git.v4/plumbing/object"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/cli-experimental/internal/pkg/client"
+	"sigs.k8s.io/cli-experimental/internal/pkg/clik8s"
+	"sigs.k8s.io/kustomize/pkg/inventory"
+)
+
+// Prune prunes directories
+type Prune struct {
+	DynamicClient client.Client
+	Out           io.Writer
+	Resources     clik8s.ResourcePruneConfigs
+	Commit        *object.Commit
+}
+
+// Result contains the Prune Result
+type Result struct {
+	Resources clik8s.ResourceConfigs
+}
+
+// Do executes the prune
+func (o *Prune) Do() (Result, error) {
+	if o.Resources == nil {
+		return Result{}, nil
+	}
+	fmt.Fprintf(o.Out, "Doing `cli-experimental prune`\n")
+
+	u := (*unstructured.Unstructured)(o.Resources)
+	annotation := u.GetAnnotations()
+	_, ok := annotation[inventory.InventoryAnnotation]
+	if !ok {
+		return Result{}, nil
+	}
+
+	obj := u.DeepCopy()
+	err := o.DynamicClient.Get(context.Background(),
+		types.NamespacedName{Namespace: u.GetNamespace(), Name: u.GetName()}, obj)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// no prune configmap set by apply, therefor we can't prune anything
+			return Result{}, nil
+		}
+		fmt.Fprintf(os.Stderr, "retrieving current configuration of %s from server for %v", u.GetName(), err)
+		return Result{}, err
+	}
+	obj, results, err := o.runPrune(obj)
+	if err != nil {
+		return Result{}, err
+	}
+
+	err = o.DynamicClient.Apply(context.Background(), obj)
+	if err != nil {
+		return Result{}, err
+	}
+
+	return Result{Resources: results}, nil
+}
+
+// runPrune deletes the obsolete objects.
+// The obsolete objects is derived by parsing
+// an Inventory annotation, which is defined in
+// Kustomize.
+//     https://github.com/kubernetes-sigs/kustomize/tree/master/pkg/inventory
+// This is based on the KEP
+//     https://github.com/kubernetes/enhancements/pull/810
+func (o *Prune) runPrune(obj *unstructured.Unstructured) (
+	*unstructured.Unstructured, []*unstructured.Unstructured, error) {
+	var results []*unstructured.Unstructured
+	annotations := obj.GetAnnotations()
+	inv := inventory.NewInventory()
+	inv.LoadFromAnnotation(annotations)
+	items := inv.Prune()
+	for _, item := range items {
+		gvk := schema.GroupVersionKind{
+			Group:   item.Group,
+			Version: item.Version,
+			Kind:    item.Kind,
+		}
+		u, err := o.deleteObject(gvk, item.Namespace, item.Name)
+		if err != nil {
+			return nil, nil, err
+		}
+		if u != nil {
+			results = append(results, u)
+		}
+	}
+	inv.UpdateAnnotations(annotations)
+	obj.SetAnnotations(annotations)
+	return obj, results, nil
+}
+
+func (o *Prune) deleteObject(gvk schema.GroupVersionKind, ns, nm string) (*unstructured.Unstructured, error) {
```

Comment:
how is this function different than the other one?

Reply:
@pwittrock Which one do you mean? If you mean the `deleteObject` function from the `delete` package, they are the same.